internal/k8stest/k8s_collector.go

```diff
@@ -53,15 +56,49 @@ func WaitForCollectorToStart(t *testing.T, client *dynamic.DynamicClient, podNam
 	podGVR := schema.GroupVersionResource{Version: "v1", Resource: "pods"}
 	listOptions := metav1.ListOptions{LabelSelector: SelectorFromMap(podLabels).String()}
 	podTimeoutMinutes := 3
-	var podPhase string
-	require.Eventually(t, func() bool {
+	t.Logf("waiting for collector pods to be ready")
+	require.Eventuallyf(t, func() bool {
 		list, err := client.Resource(podGVR).Namespace(podNamespace).List(context.Background(), listOptions)
 		require.NoError(t, err, "failed to list collector pods")
-		if len(list.Items) == 0 {
+		podsNotReady := len(list.Items)
+		if podsNotReady == 0 {
+			t.Log("did not find collector pods")
 			return false
 		}
-		podPhase = list.Items[0].Object["status"].(map[string]any)["phase"].(string)
-		return podPhase == "Running"
-	}, time.Duration(podTimeoutMinutes)*time.Minute, 50*time.Millisecond,
-		"collector pod haven't started within %d minutes, latest pod phase is %s", podTimeoutMinutes, podPhase)
+
+		var pods v1.PodList
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(list.UnstructuredContent(), &pods)
+		require.NoError(t, err, "failed to convert unstructured to podList")
+
+		for _, pod := range pods.Items {
+			podReady := false
+			if pod.Status.Phase != v1.PodRunning {
+				t.Logf("pod %v is not running, current phase: %v", pod.Name, pod.Status.Phase)
+				continue
+			}
+			for _, cond := range pod.Status.Conditions {
+				if cond.Type == v1.PodReady && cond.Status == v1.ConditionTrue {
+					podsNotReady--
+					podReady = true
+				}
+			}
+			// Add some debug logs for crashing pods
+			if !podReady {
+				for _, cs := range pod.Status.ContainerStatuses {
+					restartCount := cs.RestartCount
+					if restartCount > 0 && cs.LastTerminationState.Terminated != nil {
+						t.Logf("restart count = %d for container %s in pod %s, last terminated reason: %s", restartCount, cs.Name, pod.Name, cs.LastTerminationState.Terminated.Reason)
+						t.Logf("termination message: %s", cs.LastTerminationState.Terminated.Message)
+					}
+				}
+			}
+		}
+		if podsNotReady == 0 {
+			t.Logf("collector pods are ready")
+			return true
+		}
+		return false
```

Comment:
I see what we're doing here, but instead of a count, is there a way we can track that all pods are ready via return statements?  While looping through each pod, if any pod is not "ready" (whatever that check is) can we return false? If we make it through the whole list of pods then we know all the pods are "ready" and we can return true.

Reply:
It is kind of doing that already, except we don't return on the first not ready pod found. I am looping over all pods intentionally cause I wanted to see the status/debug info of all the collector pods while debugging the failed tests. I don't mind changing it to exit early on first not ready pod if you prefer that.

Reply:
If the goal is to keep cycling through the pods to print as much help info as possible I'm on board.